doc/src/manual/interfaces.md

```diff
@@ -371,3 +371,190 @@ If you are defining an array type that allows non-traditional indexing (indices
+`Broadcast.rule(S, T)` for two types `S` and `T`.
+You do not need to write a binary `rule` unless you want to establish precedence for
+two or more non-`BottomArray` types. If you do write a binary rule, you do not need to
+supply the types in both orders, as internal machinery will try both. For more detail,
+see [below](@ref writing-binary-broadcasting-rules).
+
+The actual allocation of the result array is handled by specialized implementations of `similar`:
+
+```julia
+Base.similar(f, r::Broadcast.Result{ContainerType}, As...)
+```
+
+`f` is the operation being performed and `ContainerType` signals the resulting
+container type (e.g., `Broadcast.BottomArray`, `Tuple`, etc.).
+`eltype(r)` returns the element type, and `indices(r)` the object's indices.
+`As...` is the list of input objects. You may not need to use `f` or `As...`
+unless they help you build the appropriate object; the fallback definition is
+
+```julia
+Base.similar(f, r::Broadcast.Result{BottomArray}, As...) = similar(Array{eltype(r)}, indices(r))
+```
+
+However, if needed you can specialize on any or all of these arguments.
+
+For a complete example, let's say you have created a type, `ArrayAndChar`, that stores an
+array and a single character:
+
+```jldoctest
+struct ArrayAndChar{T,N} <: AbstractArray{T,N}
+    data::Array{T,N}
+    char::Char
+end
+Base.size(A::ArrayAndChar) = size(A.data)
+Base.getindex(A::ArrayAndChar{T,N}, inds::Vararg{Int,N}) where {T,N} = A.data[inds...]
+Base.setindex!(A::ArrayAndChar{T,N}, val, inds::Vararg{Int,N}) where {T,N} = A.data[inds...] = val
+Base.showarg(io::IO, A::ArrayAndChar, toplevel) = print(io, typeof(A), " with char '", A.char, "'")
+```
+
+You might want broadcasting to preserve the `char` "metadata." First we define
+
+```jldoctest
+Broadcast.rule(::Type{AC}) where AC<:ArrayAndChar = ArrayAndChar
+```
+
+This forces us to also define a `similar` method:
+```jldoctest
+function Base.similar(f, r::Broadcast.Result{ArrayAndChar}, As...)
+    # Scan the inputs for the ArrayAndChar:
+    A = find_aac(As...)
+    # Use the char field of A to create the output
+    ArrayAndChar(similar(Array{eltype(r)}, indices(r)), A.char)
+end
+
+"`A = find_aac(As...)` returns the first ArrayAndChar among the arguments."
+find_aac(A::ArrayAndChar, B...) = A
+find_aac(A, B...) = find_aac(B...)
+```
+
+From these definitions, one obtains the following behavior:
+```jldoctest
+julia> a = ArrayAndChar([1 2; 3 4], 'x')
+2×2 ArrayAndChar{Int64,2} with char 'x':
+ 1  2
+ 3  4
+
+julia> a .+ 1
+2×2 ArrayAndChar{Int64,2} with char 'x':
+ 2  3
+ 4  5
+
+julia> a .+ [5,10]
+2×2 ArrayAndChar{Int64,2} with char 'x':
+  6   7
+ 13  14
+```
+
+Finally, it's worth noting that sometimes it's easier simply to bypass the machinery for
+computing result types and container sizes, and just do everything manually. For example,
+you can convert a `UnitRange{Int}` `rng` to a `UnitRange{BigInt}` with `big.(rng)`; the definition
```

Comment:
`rng` is generally used for "random number generator", better use `r`.

Reply:
I wanted to disambiguate it from `r::Broadcast.Result`, which I use in the same paragraph.

Reply:
`rng` tripped me up for the reason Milan gave, so another vote for something other than `rng` :).

Reply:
Maybe you can change `rng` to `r` and `r::Broadcast.Result` to `res::Broadcast.Result`?